x/stake/endblock.go

```diff
@@ -252,3 +266,84 @@ func handleMatureUnbondingDelegations(k keeper.Keeper, ctx sdk.Context) ([]types
 
 	return completed, events
 }
+
+const (
+	maxProcessedRefundCount = 50
+)
+
+func handleRefundCrossStake(ctx sdk.Context, k keeper.Keeper) sdk.Events {
+	iterator := k.IteratorAllDelegations(ctx)
+	defer iterator.Close()
+	var refundEvents sdk.Events
+	count := 0
+	bcValidators := k.GetAllValidators(ctx)
+	bcValidatorsMap := make(map[string]types.Validator, len(bcValidators))
+	for _, v := range bcValidators {
+		bcValidatorsMap[v.OperatorAddr.String()] = v
+	}
+	for ; iterator.Valid(); iterator.Next() {
+		delegation := types.MustUnmarshalDelegation(k.CDC(), iterator.Key(), iterator.Value())
+		if delegation.CrossStake {
+			result := handleMsgSideChainUndelegate(ctx, types.MsgSideChainUndelegate{
+				DelegatorAddr: delegation.DelegatorAddr,
+				ValidatorAddr: delegation.ValidatorAddr,
+				Amount:        sdk.NewCoin(k.BondDenom(ctx), delegation.GetShares().RawInt()),
+				SideChainId:   k.ScKeeper.BscSideChainId(ctx),
+			}, k)
+			refundEvents = refundEvents.AppendEvents(result.Events)
+		} else {
+			if _, exist := bcValidatorsMap[delegation.DelegatorAddr.String()]; !exist {
```

Comment:
Does this check necessary, because I can not figure a case that delegation exist while validator is gone.

Reply:
The useless code `bcValidatorsMap` is not deleted yet.

Reply:
fixed